src/systemd-sonic-generator/systemd-sonic-generator.c

```diff
@@ -0,0 +1,299 @@
+    int num_targets;
+    int found_targets;
+    char* token;
+    char* line = NULL;
+    bool first;
+    char* target_suffix;
+
+    strcpy(file_path, UNIT_FILE_PREFIX);
+    strcat(file_path, unit_file);
+
+    num_target_lines = get_target_lines(file_path, target_lines);
+    if (num_target_lines < 0)
+        return -1;
+
+    num_targets = 0;
+
+    for (int i = 0; i < num_target_lines; i++) {
+        line = target_lines[i];
+        first = true;
+
+        while ((token = strtok_r(line, "=", &line))) {
+            if (first) {
+                first = false;
+
+                if (strstr(token, "RequiredBy") != NULL) {
+                    target_suffix = ".requires";
+                }
+                else if (strstr(token, "WantedBy") != NULL) {
+                    target_suffix = ".wants";
+                }
+            }
+            else {
+                found_targets = get_install_targets_from_line(token, target_suffix, &targets[num_targets]);
+                if (found_targets < 0)
+                    return -1;
+
+                num_targets += found_targets;
+            }
+        }
+        free(target_lines[i]);
+    }
+    return num_targets;
+}
+
+
+static int get_unit_files(char* unit_files[]) {
+    /***
+    Reads a list of unit files to be installed from /etc/sonic/generated_services.conf
+    ***/
+    FILE *fp;
+    char *line = NULL;
+    size_t len = 0;
+    ssize_t read;
+
+    fp = fopen(CONFIG_FILE, "r");
+
+    if (fp == NULL) {
+        exit(EXIT_FAILURE);
+    }
+
+    int num_unit_files = 0;
+
+    while ((read = getline(&line, &len, fp)) != -1) {
+        strip_trailing_newline(line);
+        unit_files[num_unit_files] = strdup(line);
+        num_unit_files++;
+    }
+
+    free(line);
+
+    fclose(fp);
+
+    return num_unit_files;
+}
+
+
+static int install_unit_file(char* unit_file, char* target, char* install_dir) {
+    /***
+    Creates a symlink for a unit file installation
+
+    For a given unit file and target directory,
+    create the appropriate symlink in the target directory
+    to enable the unit and have it started by Systemd
+    ***/
+    char final_install_dir[PATH_MAX];
+    char src_path[PATH_MAX];
+    char dest_path[PATH_MAX];
+    struct stat st;
+    int r;
+
+    assert(unit_file);
+    assert(target);
+
+    strcpy(final_install_dir, install_dir);
+    strcat(final_install_dir, target);
+
+    strcpy(src_path, UNIT_FILE_PREFIX);
+    strcat(src_path, unit_file);
+
+    if (stat(final_install_dir, &st) == -1) {
+        // If doesn't exist, create
+        r = mkdir(final_install_dir, 0755);
+        if (r == -1) {
+            fputs("Unable to create target directory\n", stderr);
+            return -1;
+        }
+    }
+    else if (S_ISREG(st.st_mode)) {
+        // If is regular file, remove and create
+        r = remove(final_install_dir);
+        if (r == -1) {
+            fputs("Unable to remove file with same name as target directory\n", stderr);
```

Comment:
do those message went to syslog?

Reply:
print out the final_install_dir in the message.

Reply:
I don't think syslog is running yet, since this program runs before Systemd starts

Reply:
I can manually write to /var/log/syslog, would that work? Otherwise I think these messages just go to wherever stderr is